linear-algebra/src/main/scala/scalan/linalgebra/Matrices.scala

```diff
@@ -89,112 +101,234 @@ trait Matrices extends Vectors { self: MatricesDsl =>
+      }
     }
 
-    def +^^(other: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
-      companion((rows zip other.rows).flatMap { case Pair(v1, v2) => (v1 +^ v2).items }, numColumns)
+    @OverloadId("matrix")
+    def +^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
+      matrix match {
+        case DenseFlatMatrix(rmValues1, _) =>
+          DenseFlatMatrix((DenseVector(rmValues) +^ DenseVector(rmValues1)).items, numColumns)
+        case CompoundMatrix(rows1, _) =>
+          DenseFlatMatrix((rows zip rows1).flatMap { case Pair(v1, v2) => (v1 +^ v2).items }, numColumns)
+        case ConstMatrix(value, _, _) =>
+          DenseFlatMatrix((DenseVector(rmValues) +^ value).items, numColumns)
+        case DiagonalMatrix(diagonal) =>
+          val width = diagonal.length
+          val mainDiagonalIndices = Collection.indexRange(width).map(i => i * width + i)
+          val newValues = (diagonal zip rmValues(mainDiagonalIndices)).map { case Pair(d, v) => d + v }
+          DenseFlatMatrix(rmValues.updateMany(mainDiagonalIndices, newValues), numColumns)
+        case ConstDiagonalMatrix(diagonalValue, width) =>
+          val mainDiagonalIndices = Collection.indexRange(width).map(i => i * width + i)
+          val newValues = rmValues(mainDiagonalIndices).map(v => diagonalValue + v)
+          DenseFlatMatrix(rmValues.updateMany(mainDiagonalIndices, newValues), numColumns)
+        // TODO: check case _
+        case _ => !!!("matcher for @matrix argument in DenseFlatMatrix.+^^(matrix: Matr[T]) is not specified.")
+      }
     }
 
     @OverloadId("matrix")
-    def *^^(other: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
-      other match {
+    def *^^(matrix: Matr[T])(implicit n: Numeric[T]): Matr[T] = {
+      matrix match {
         case DenseFlatMatrix(rmValues1, _) =>
-          companion((rmValues zip rmValues1).map { case Pair(v1, v2) => v1 * v2 }, numColumns)
+          DenseFlatMatrix((DenseVector(rmValues) *^ DenseVector(rmValues1)).items, numColumns)
         case CompoundMatrix(rows1, _) =>
           CompoundMatrix((rows zip rows1).map { case Pair(v1, v2) => v1 *^ v2 }, numColumns)
-        case _ =>
-          other *^^ self
+        case ConstMatrix(value, _, _) =>
+          DenseFlatMatrix((DenseVector(rmValues) *^ value).items, numColumns)
+        case DiagonalMatrix(diagonalValues) =>
+          val width = diagonalValues.length
+          val mainDiagonalIndices = Collection.indexRange(width).map(i => i * width + i)
+          DiagonalMatrix((DenseVector(rmValues(mainDiagonalIndices)) *^ DenseVector(diagonalValues)).items)
+        case ConstDiagonalMatrix(diagonalValue, width) =>
+          val mainDiagonalIndices = Collection.indexRange(width).map(i => i * width + i)
+          DiagonalMatrix((DenseVector(rmValues(mainDiagonalIndices)) *^ diagonalValue).items)
+        // TODO: check case _
+        case _ => !!!("matcher for @matrix argument in DenseFlatMatrix.*^^(matrix: Matr[T]) is not specified.")
```

Comment:
Operator is argument-specific here. Different representations of rhs matrix should yield different code (and different matrix representation as a result). Not sure how it could be achieved otherwise, though it'd be cool to reduce the matchers code.